integration-tests/suite_test.go

```diff
@@ -0,0 +1,250 @@
+}
+
+type ContainerRequest = testcontainers.ContainerRequest
+
+// RunContainer runs a container, capturing its logs as artifacts, and stopping
+// the container when the test finishes.
+func (s *Suite) RunContainer(t *testing.T, name string, captureLogs bool, req ContainerRequest) *Container {
+	t.Helper()
+
+	ctx := s.Context(t)
+
+	req.Name = fmt.Sprintf("%s-%s", s.Name, name)
+	req.AutoRemove = false
+	req.SkipReaper = disableReaper
+
+	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+	require.NoError(t, err)
+
+	logs := &logConsumer{}
+	container.FollowOutput(logs)
+
+	require.NoError(t, container.StartLogProducer(ctx))
+
+	t.Cleanup(func() {
+		if err := container.StopLogProducer(); err != nil {
+			t.Logf("failed to stop log producer: %v\n", err)
+		}
+
+		if captureLogs {
+			s.CaptureArtifact(
+				fmt.Sprintf("%s.log", name),
+				logs.bytes(),
+			)
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		state, err := container.State(ctx)
+		if err != nil {
+			t.Logf("failed to get container state (%s): %v\n", req.Name, err)
+			return
+		}
+
+		if !state.Running {
+			return
+		}
+
+		if err := container.Terminate(ctx); err != nil {
+			t.Logf("failed to terminate container (%s): %v\n", req.Name, err)
+		}
+	})
+
+	hostIP, err := container.Host(ctx)
+	require.NoError(t, err)
+
+	containerIP, err := container.ContainerIP(ctx)
+	require.NoError(t, err)
+
+	hostMappedPorts := make(map[nat.Port]int, len(req.ExposedPorts))
+	for _, portString := range req.ExposedPorts {
+		port := nat.Port(portString)
+		hostPort, err := container.MappedPort(ctx, port)
+		require.NoError(t, err)
+		hostMappedPorts[port] = hostPort.Int()
+	}
+
+	return &Container{
+		Name:        req.Name,
+		Container:   container,
+		HostIP:      hostIP,
+		ContainerIP: containerIP,
+		MappedPorts: hostMappedPorts,
+	}
+}
+
+// Context returns a context that will be canceled when the test finishes.
+func (s *Suite) Context(t *testing.T) context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+	return ctx
+}
+
+// CaptureArtifact stores the given data to be written to disk when the test
+// finishes.
+func (s *Suite) CaptureArtifact(name string, data []byte) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.artifacts[name] = data
+}
+
+func (s *Suite) cleanup(t *testing.T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if outputDir == "" {
+		return
+	}
+
+	for name, data := range s.artifacts {
+		if err := os.WriteFile(filepath.Join(outputDir, name), data, 0660); err != nil {
+			t.Logf("failed to write artifact %s: %v", name, err)
+		}
+	}
+}
+
+// Volume returns a Docker volume that can be used to share files between
+// containers. You can also add files from the host using WriteFile. The
+// volume will be deleted when the test finishes.
+func (s *Suite) Volume(t *testing.T) *Volume {
```

Comment:
Totally non-blocking / optional: Did you consider docker compose for setting up some of this? 

I think a compose file could get us the docker network, the volume, and the pause container pretty easily, and we could launch additional containers from golang at that point. The docker compose config could also be customizable with something like [env variable substitution](https://docs.docker.com/compose/environment-variables/), an [override file](https://docs.docker.com/compose/extends/#multiple-compose-files), or maybe by templating the compose file.

Feel free to ignore this, since it would basically be a significant rewrite (and I'm probably missing context on broader integration test concerns/plans). Also, compose may not ultimately be better. I do think maintaining compose yaml for setup/teardown would be easier to grok. Although, we still need go code to inspect container logs to and wait for startup and such.

Reply:
Great point! To be totally honest, it's been several years since I've used compose - so it isn't in my frequent grab-bag of tools 😄 

I have a vague preference for keeping as much as possible in Go, because I find that this kind of test suite can become quite hard to reason about once you're shelling-out to different tools and templating stuff. That said, I appreciate that some of the helper functions here aren't the easiest to follow!